nixos/modules/system/etc/setup-etc.pl

```diff
@@ -1,52 +1,64 @@
+#! @perl@/bin/perl
+
 use strict;
-use File::Find;
-use File::Copy;
-use File::Path;
-use File::Basename;
-use File::Slurp;
+use warnings;
+
+use File::Basename qw(dirname);
+use File::Copy qw(copy);
+use File::Find qw(find);
+use File::Path qw(make_path rmtree);
+use File::Slurp qw(read_file write_file);
+
+## no critic(CodeLayout::ProhibitParensWithBuiltins)
+## no critic(ErrorHandling::RequireCarping)
+## no critic(InputOutput::RequireCheckedSyscalls, InputOutput::RequireBracedFileHandleWithPrint, InputOutput::RequireBriefOpen)
+## no critic(RegularExpressions::ProhibitEscapedMetacharacters)
+## no critic(ValuesAndExpressions::ProhibitNoisyQuotes, ValuesAndExpressions::ProhibitMagicNumbers, ValuesAndExpressions::ProhibitEmptyQuotes, ValuesAndExpressions::ProhibitInterpolationOfLiterals)
 
-my $etc = $ARGV[0] or die;
+my $etc = $ARGV[0] or die("This script must be called with an argument which expresses the new /etc directory in the nix store");
 my $static = "/etc/static";
 
-sub atomicSymlink {
+# Create a symlink atomically from argument 2 -> argument 1
+sub atomic_symlink {
     my ($source, $target) = @_;
     my $tmp = "$target.tmp";
-    unlink $tmp;
-    symlink $source, $tmp or return 0;
-    rename $tmp, $target or return 0;
+    unlink($tmp);
+    symlink($source, $tmp) or return 0;
+    rename($tmp, $target) or return 0;
     return 1;
 }
 
 
 # Atomically update /etc/static to point at the etc files of the
 # current configuration.
-atomicSymlink $etc, $static or die;
+atomic_symlink($etc, $static) or die("Failed to create symlink for /etc");
 
 # Returns 1 if the argument points to the files in /etc/static.  That
 # means either argument is a symlink to a file in /etc/static or a
 # directory with all children being static.
-sub isStatic {
-    my $path = shift;
+sub is_static {
+    my $path = @_;
 
     if (-l $path) {
-        my $target = readlink $path;
-        return substr($target, 0, length "/etc/static/") eq "/etc/static/";
+        my $target = readlink($path);
+        return substr($target, 0, length($static)) eq $static;
     }
 
     if (-d $path) {
-        opendir DIR, "$path" or return 0;
-        my @names = readdir DIR or die;
-        closedir DIR;
+        opendir(my $dir, "$path") or return 0;
+        my @names = readdir($dir) or die("Failed reading directory `$dir`");
+        closedir($dir);
 
         foreach my $name (@names) {
-            next if $name eq "." || $name eq "..";
-            unless (isStatic("$path/$name")) {
+            if ($name eq "." || $name eq "..") {
+                next;
+            }
+            if (not (is_static("$path/$name"))) {
                 return 0;
             }
         }
```

Comment:
- Avoid `return 0` as it is true in list context.

Note: The same logic could be implemented as:
```perl
return if grep { not is_static("$path/$_") } @names;
```

Reply:
Or:
```
foreach my $name (@names) {
    return unless is_static("$path/name")
}
```

Reply:
so simply returning without value would be enough here?